### PR TITLE
Lexer errors

### PR DESCRIPTION
**Pull request by JHodorHodor**

### CODE REVIEW

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
-        }
+    class LexerError(val start: Location, val end: Location?, val context: List<String>, val errorSegment: String) : Diagnostic() {
+
+        override fun isError(): Boolean = true
+
+        override fun toString() = StringBuilder()
+            .append("Unable to match token at location $start - ${end ?: "eof"}.\n")
+            .append("\t\t${context.joinToString("")}-->$errorSegment<---")
+            .toString()
     }
```
Comment (rkilar) [inline app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:16]: Not sure about the context being a list of strings if we join them anyway. Maybe we can just make it a single string. Also not sure about the context containing the 3 last tokens. Having the context be the beginning of the line containing the error would make it more readable. We could also include the rest of the line.

**Review by rkilar — CHANGES_REQUESTED** [review]
The code looks good. Maybe we can change what we store as context for the error. Storing the content of the line containing the error would make it more readable.

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt (hunk 1 above)
Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:16]: I thought about this being a line, but the user already receives the line number, so to give him more compact information I chose the last 3 tokens.

File: app/src/test/kotlin/compiler/lexer/LexerTest.kt [hunk 2]
```diff
@@ -129,24 +141,134 @@ class LexerTest {
...
-        val lexer = Lexer<Unit>(listOf(Pair(dfa, Unit)))
+        val lexer = Lexer<Unit>(listOf(Pair(dfa, Unit)), TestDiagnostics())
 
         lexer.process(input).count()
 
         assertContains(input.flushed, 2)
         assertContains(input.flushed, 4)
     }
+
+    @Test fun `an input matching to no tokens results only in an error`() {
```
[the remark this answers is not in the document]
Comment (JHodorHodor) [reply app/src/test/kotlin/compiler/lexer/LexerTest.kt:155]: Already in this file there are older tests with names not starting with `test`... I tried to keep it consistent in the first sprint, but it turns out that I failed. Maybe at some point, we could try to clean this up (and discuss a consistent way of naming).

File: app/src/main/kotlin/compiler/lexer/Lexer.kt [hunk 3]
```diff
@@ -2,23 +2,42 @@ package compiler.lexer
...
+            context.addLast(content)
+            if (context.size > contextErrorLength) context.removeFirst()
+        }
+
         while (input.hasNext()) {
             input.flush() // Tell the input source that previous characters can be discarded.
 
+            if (isErrorSegment) {
+                errorSegment.append(input.next())
+            }
```
[the remark this answers is not in the document]
Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/lexer/Lexer.kt]: Done

File: app/src/test/kotlin/compiler/lexer/LexerTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkilar) [reply app/src/test/kotlin/compiler/lexer/LexerTest.kt:155]: I'm not so sure about this convention myself.

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt (hunk 1 above)
Comment (rkilar) [reply app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:16]: It guess it's fine. My concern is that if the error is at the beginning of the line, then one of the tokens might be a newline and the error wouldn't look all that good.

**Review by rkilar — APPROVED** [review]

**Pull request merged**